topdown/graphql.go

```diff
@@ -0,0 +1,293 @@
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// Parses a GraphQL schema, and returns only the Schema object.
+// Used in validation of queries.
+// NOTE(philipc): The error type here is a gqlerror.Error struct,
+// which requires us to treat it specially, since the returned value
+// will *always* be non-null.
+// See: https://staticcheck.io/docs/checks#SA4023 for details.
+func loadSchema(schema string) (*gqlast.Schema, error) {
+	loadedSchema, err := gqltop.LoadSchema(&gqlast.Source{Input: schema})
+	if err != nil {
+		return nil, err
+	}
+	return loadedSchema, nil
+}
+
+// Parses a GraphQL schema, and returns the GraphQL AST for the schema.
+// NOTE(philipc): The error type here is a gqlerror.Error struct,
+// which requires us to treat it specially, since the returned value
+// will *always* be non-null.
+// See: https://staticcheck.io/docs/checks#SA4023 for details.
+func parseSchema(schema string) (*gqlast.SchemaDocument, error) {
+	schemaAST, err := gqlparser.ParseSchema(&gqlast.Source{Input: schema})
+	if err != nil {
+		return nil, err
+	}
+	return schemaAST, nil
+}
+
+// Parses a GraphQL query, and returns the GraphQL AST for the query.
+// NOTE(philipc): The error type here is a gqlerror.Error struct,
+// which requires us to treat it specially, since the returned value
+// will *always* be non-null.
+// See: https://staticcheck.io/docs/checks#SA4023 for details.
+func parseQuery(builtinName string, query string) (*gqlast.QueryDocument, error) {
+	queryAST, err := gqlparser.ParseQuery(&gqlast.Source{Input: query})
+	if err != nil {
+		return nil, &Error{
+			Code:     BuiltinErr,
+			Message:  fmt.Sprintf("%s error: %s", builtinName, err.Message),
+			Location: ast.NewLocation([]byte(builtinName), "input", err.Locations[0].Line, err.Locations[0].Column),
+		}
+	}
+	return queryAST, nil
+}
+
+// Validates a GraphQL query against a schema, and returns an error list.
+// NOTE(philipc): The error type here is a gqlerror.Error struct,
+// which requires us to treat it specially, since the returned value
+// will *always* be non-null.
+// See: https://staticcheck.io/docs/checks#SA4023 for details.
+func validateQuery(schema *gqlast.Schema, query *gqlast.QueryDocument) error {
+	// Validate the query against the schema, erroring if there's an issue.
+	err := gqlvalidator.Validate(schema, query)
+	if err != nil {
+		return errGraphQLErrorList(ast.GraphQLParse.Name, err)
+	}
+	return nil
+}
+
+// Generate a useful error message from a GraphQL error list.
+// We return an ast.Errors structure here because it is the closest match.
+func errGraphQLErrorList(builtinName string, errList gqlerror.List) error {
+	errs := make(ast.Errors, len(errList))
+	for _, e := range errList {
+		fmt.Println(e)
+		errs = append(errs, &ast.Error{
+			Code:     BuiltinErr,
+			Message:  fmt.Sprintf("GraphQL Error: %s", e.Message),
+			Location: ast.NewLocation([]byte(builtinName), "", e.Locations[0].Line, e.Locations[0].Column),
```

Comment:
That location, does it make sense from rego's perspective? Also I'd have expected some helper, newbuiltinerr or something like that.